streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -632,29 +749,60 @@ public void onAssignment(final Assignment assignment) {
         Collections.sort(partitions, PARTITION_COMPARATOR);
 
         final AssignmentInfo info = AssignmentInfo.decode(assignment.userData());
-        final int usedVersion = info.version();
+        final int receivedAssignmentMetadataVersion = info.version();
+        final int supportedVersion = info.latestSupportedVersion();
+
+        if (usedSubscriptionMetadataVersion > receivedAssignmentMetadataVersion
+            && receivedAssignmentMetadataVersion >= EARLIEST_PROBEABLE_VERSION) {
```

Comment:
`receivedAssignmentMetadataVersion >= EARLIEST_PROBEABLE_VERSION` should be guaranteed at the server side as always right? If that is true, I'd suggest we refactor it as:

```
if (usedSubscriptionMetadataVersion > receivedAssignmentMetadataVersion) {
    if (receivedAssignmentMetadataVersion < EARLIEST_PROBEABLE_VERSION) {
        // throw illegal state exception.
    }

    // .. below logic
}
```

So that we can detect potential bugs.

Reply:
This code is actually correct. If I change a suggested, it break for the manual upgrade path from version 1/2 to version 3. For this case, a received version can be smaller than the used-subscription-version, and smaller than EARLIEST_PROBEABLE_VERSION.

Reply:
Ah you're right. My understanding is that to manual upgrade from version 1/2 to version 3, we set `upgrade.from` config accordingly, so first rebalance everyone use version 1/2 in subscriptionInfo and AssignmentInfo; then in second rebalance someone send subscriptionInfo with version 3, and someone send with version 1/2 (they have not bounced yet), so assignmentInfo with version 1/2 are sent back again.